File: 1-two-sum/two-sum.cpp
/*  class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {

        for(int i=0; i<nums.size(); i++){
            for(int j=i+1; j<nums.size(); j++){
                if(nums[i] + nums[j] == target) return {i, j};
            }
        }
         return {};
    }

};  */

/* class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mp;

        for (int i = 0; i < nums.size(); i++) {
            int need = target - nums[i];

            if (mp.count(need)) {
                return {mp[need], i};
            }

            mp[nums[i]] = i;
        }

        return {};
    }
}; */

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {

        vector<pair<int, int>> arr;

        for(int i = 0; i < nums.size(); i++) {
            arr.push_back({nums[i], i});
        }

        sort(arr.begin(), arr.end());

        int i = 0;
        int j = arr.size() - 1;

        while(i < j) {

            int sum = arr[i].first + arr[j].first;

            if(sum == target) {
                return {arr[i].second, arr[j].second};
            }
            else if(sum < target) {
                i++;
            }
            else {
                j--;
            }
        }

        return {};
    }
};

File: 1-two-sum/two-sum.js
/* var twoSum = function(nums, target) {

    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {

            if (nums[i] + nums[j] === target) {
                return [i, j];
            }

        }
    }

}; */

var twoSum = function(nums, target) {

    let arr = nums.map((value, index) => [value, index]);

    arr.sort((a, b) => a[0] - b[0]);

    let i = 0;
    let j = arr.length - 1;

    while (i < j) {

        let sum = arr[i][0] + arr[j][0];

        if (sum === target) {
            return [arr[i][1], arr[j][1]];
        }
        else if (sum < target) {
            i++;
        }
        else {
            j--;
        }
    }

    return [];
};

